### PR TITLE
Let's make sure we are not copying improper rows

### PR DESCRIPTION
**Pull request by photodude**
Insurance against cases where the records should not be copied. 
(These shouldn't happen, but if they do...)
- If it's the same record, there is nothing to do
- Don't copy over if the `$to` record has slaveusers
- Don't copy over if the `$to` subscription has slavesubs_ids

### CODE REVIEW

**Comment by nikosdion** [comment]
I can assure you that $to->slaveusers and $to->slavesubs_ids will always return null since there are no such fields in the table and, by extension, in the class. They will end up doing $to->getState('slaveusers', null, 'raw') which returns null since there's no such state variable set. I'm not sure what isset() will do. I suppose that since we don't have a magic __isset() method in DataModel trying to use isset() with any magic variable (be it a field or a state variable) would return false. I never tried it, I never say the reason to: we already have methods to check for the existence of fields.

Anyway. I think you meant to access $to->params instead?

**Comment by photodude** [comment]
You are correct I intended to have these check against the $to->params items. 
My understanding of [isset()](http://php.net/manual/en/function.isset.php) is it Returns TRUE if var exists and has value other than NULL, FALSE otherwise. Which off hand seems like the appropriate choice for these checks. Do you think  another "method to check for the existence of fields" is more appropriate in this case?
I'll take a look at this again on Wednesday.

**Comment by nikosdion** [comment]
No. isset() returns true if the variable / property exists, false otherwise. What you are mistaking it with is  empty().

However, isset() won't work with DataModel properties which give you access to the record's fields. If you look at the code you will see that these properties are not defined in the model class. They are "virtual" properties. When you access $subscription->akeebasubs_subscription_id you will see that this property doesn't exist. PHP sees that too and calls the magic __get() method in DataModel. This method tries to figure out if a record field or model state variable by the same name exists and returns its value. However, the operating phrase here is "this property doesn't exist". Since it doesn't exist isset() will return false.

PHP allows objects to implement a magic __isset() method as well. If isset() can't find a property it calls the object's magic __isset() method which returns true if the property should be considered as existing (even though it's not there, e.g. a known magic property), false otherwise. We don't have that in FOF's Model or DataModel classes, therefore trying isset() against a magic property always returns false.

TL;DR: Don't use isset() against Model's and DataModel's magic properties (record fields, model state variables). It won't work.

**Comment by wilsonge** [comment]
isset returns false if there is a var that is null.

**Comment by nikosdion** [comment]
Still, as I said:

> PHP allows objects to implement a magic __isset() method as well. If isset() can't find a property it calls the object's magic __isset() method which returns true if the property should be considered as existing (even though it's not there, e.g. a known magic property), false otherwise. We don't have that in FOF's Model or DataModel classes, therefore trying isset() against a magic property always returns false.

So, really, don't use isset().

**Comment by photodude** [comment]
When I had wrote this I had [compared `empty()` and `isset()`](https://www.virendrachandak.com/techtalk/php-isset-vs-empty-vs-is_null/), From the table of Boolean values I had reviewed; It seemed that `isset()` might have been the best choice for the checks. What I had failed to consider is we are working with **DataModel Objects and DataModel properties** and not just variables. If I understand it right; when `isset()` is used on a DataModel property it detects if the property has been initialized, and will return false since all DataModel properties are always considered set even when empty.

So I agree `isset()` shouldn't be used.

(on a side note: this is consistent with information I recently read in [PHP Cookbook: Solutions & Examples for PHP Programmers](https://books.google.com/books?id=UubtAwAAQBAJ&pg=PA203&lpg=PA203&dq=isset()+won%27t+work+with+Data+Model+properties&source=bl&ots=onLTZo6YKl&sig=NmBPTXE9VfBQpQPMEz8rhRWDdaU&hl=en&sa=X&ei=gfyHVYSOB8WuyAS09YLQDw&ved=0CEoQ6AEwBg#v=onepage&q=isset()%20won't%20work%20with%20Data%20Model%20properties&f=false) )

**Comment by nikosdion** [comment]
> it detects if the property has been initialized, and will return false since all DataModel properties are always considered set even when empty.

No. There are no public variables in the object. When you access `$sub->params` there is no `public $params` in the object. DataModel's magic `__get` method is called which executes `return $this->getFieldValue('params')`. Since there is no concrete variable in the object `isset($this->params)` will always return false no matter what the result of `$this->params` is (empty or not, it doesn't matter). This behavior is correct since `isset()` is supposed to tell you if a concrete variable is set. Magic getters do not count as concrete variables since whether they can handle a variable name or not is not known until their execution is completed.
